proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+# Introduce External Calls
+
+* Proposal: [FIP-0003](0003-external-calls.md)
+* Author: [Daniel Hails](https://github.com/djrhails) & [Alexander Harkness](https://github.com/bearbin) 
+* Review Manager:  [Susan Eisenbach](https://github.com/SusanEisenbach)
+* Status: **Awaiting review**
+* Issue label: [0003-external-calls](https://github.com/franklinsch/flint/issues?q=is%3Aopen+is%3Aissue+label%3A0003-external-calls)
+
+## Introduction
+
+Contracts can be created "from outside" via Ethereum transactions or from within Flint Contracts. They contain persistent data in state variables and functions that can modify these variables. Calling a function on a different contract (instance) will perform an EVM Function call and thus switch the context such that state variables in the old context are inaccessible.
+
+## Motivation
+Calls to untrusted contracts can introduce several unexpected risks and errors. External calls may execute malicious code in that contract _or any other contract_ that it depends upon. As such, **every** external call should be treated as a security risk.
+
+However external calls are necessary to accomplish key features of smart contracts, including:
+- Paying other users
+- Interacting with other Contracts e.g. Tokens or Wallets
+
+There have been several published best practice guidelines for programming with external calls ([Consensys Recommendations](https://consensys.github.io/smart-contract-best-practices/recommendations/#favor-pull-over-push-for-external-calls), [OpenZeppelin](http://openzeppelin.org/), [Solium Security](https://github.com/duaraghav8/solium-plugin-security), [Mythril](https://github.com/ConsenSys/mythril), [Solcheck](https://github.com/federicobond/solcheck)). This proposal will integrate best practice into the language design. Below are causes for concern with external calls.
+
+#### 1. Contracts are untrustworthy by default
+```javascript
+// bad
+Bank.withdraw(100); // Unclear whether trusted or untrusted
+
+function makeWithdrawal(uint amount) { // Isn't clear that this function is potentially unsafe
+    Bank.withdraw(amount);
+}
+
+// good
+UntrustedBank.withdraw(100); // untrusted external call
+TrustedBank.withdraw(100); // external but trusted bank contract maintained by XYZ Corp
+
+function makeUntrustedWithdrawal(uint amount) {
+    UntrustedBank.withdraw(amount);
+}
+```
+
+#### 2. External calls have arbitrary code execution
+Examples of problems that can arise in Solidity code include:
+##### Race Conditions
+Whether using raw calls (of the form `someAddress.call()`) or contract calls (of the form `ExternalContract.someMethod()`), malicious code might execute. Even if ExternalContract is not malicious, malicious code can be executed by any contracts it calls.
+One particular danger is malicious code may hijack the control flow, leading to race conditions.
```

Comment:
Race conditions are strictly a subset of reentrancy problems, so I don't think the two areas need to be mentioned separately. In any case, the concept of 'hijacking the control flow' really needs a bit more explanation.